modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -370,38 +397,32 @@ private void updateLeaseBatchInternal() {
-                        // stale leases, if watch processing was delayed for some reason. It is ok: negotiated lease is guaranteed to be
-                        // already written to meta storage before negotiation begins, and in this case its start time would be
-                        // greater than lease's.
-                        assert negotiatedLease.getStartTime().longValue() >= lease.getStartTime().longValue()
-                                : format("Can't publish the lease that was not negotiated [groupId={}, startTime={}, "
+                            // Lease information is taken from lease tracker, where it appears on meta storage watch updates, so it can
+                            // contain stale leases, if watch processing was delayed for some reason. It is ok: negotiated lease is
+                            // guaranteed to be already written to meta storage before negotiation begins, and in this case its start time
+                            // would be greater than lease's.
+                            assert negotiatedLease.getStartTime().longValue() >= lease.getStartTime().longValue()
+                                    : format("Can't publish the lease that was not negotiated [groupId={}, startTime={}, "
                                     + "agreementLeaseStartTime={}].", grpId, lease.getStartTime(), agreement.getLease().getStartTime());
 
-                        publishLease(grpId, negotiatedLease, renewedLeases);
-
-                        continue;
-                    } else if (agreement.isDeclined()) {
-                        // Here we initiate negotiations for UNDEFINED_AGREEMENT and retry them on newly started active actor as well.
-                        ClusterNode candidate = nextLeaseHolder(assignments, grpId, agreement.getRedirectTo());
+                            publishLease(grpId, negotiatedLease, renewedLeases);
 
-                        if (candidate == null) {
-                            leaseUpdateStatistics.onLeaseWithoutCandidate();
+                            continue;
+                        } else if (agreement.isDeclined()) {
+                            // Here we initiate negotiations for UNDEFINED_AGREEMENT and retry them on newly started active actor as well.
+                            chooseCandidateAndCreateNewLease(grpId, lease, agreement, assignments, renewedLeases, toBeNegotiated);
 
                             continue;
                         }
-
-                        // New lease is granted.
-                        writeNewLease(grpId, candidate, renewedLeases);
-
-                        boolean force = Objects.equals(lease.getLeaseholder(), candidate.name());
-
-                        toBeNegotiated.put(grpId, force);
+                    } else {
+                        // If the lease was denied, create the new one.
+                        chooseCandidateAndCreateNewLease(grpId, lease, agreement, assignments, renewedLeases, toBeNegotiated);
```

Comment:
Here we have the same code in two condition's branches. You can make it once.

Reply:
refactored